### DISCUSSION TITLE
Welcome to BigBagKbdTrixPKL Discussions!

### DISCUSSION THREAD

**Discussion by DreymaR** [DreymaR/BigBagKbdTrixPKL#19] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ixemul** [comment]
Hello!

First, thank you for putting all this together, it has been a big help to me.  I have a question:
How might I go about making the space bar work like a shift key if I hold it down?  I have this set up on my desktop keyboard, but on my laptop (where I'm using EPKL) I am always typing a lot of spaces when I mean to be raising to upper case.  Any thoughts?

Thanks, I'm hoping you might care because I do ;)

**Reply by DreymaR** [reply]
You can do it. In your layout, add this mapping:
`QWSPC = SPACE/LShift VirtualKey`

HOWEVER, I strongly advise against it! You'll have insane timing issues with using Space as a Tap-or-Mod key. You could fiddle with the timing settings in your EPKL settings file to see if you like the result, but I don't. When I use Shift I want a Sticky Shift instead, that I can tap and let go of before I type the shifted character. That's far more ergonomic. And since Space is the most common character by far (about 20% of key presses in text!) it's the very last key I'd want to have timing issues with.

As it stands today, it seems to me that QMK solves dual-role timing and cancels etc better than EPKL does. I have no immediate ideas about how to improve it. But at any rate I can't see how to make a ToM Space/Shift key work perfectly for me, regardless of implementation, for the reasons stated above.

**Reply by ixemul** [reply]
Thanks, yes, after a little testing I agree with you.  The way QMK handles it doesn't pose any usability issues, apart from not having key repeat on space, but the timing issues here are pretty disruptive.  

For anyone reading with the same intent I had, I'll save you a little time and describe the result -- if the timeout is too high, you'll end up not holding it long enough to get a shifted character, and if it's too short, you'll end up missing spaces between some of your words.  Not optimal!  

Anyway, thanks again, I really appreciate your quick response!  If I come up with anything I'll post back.

**Comment by DaniilHubGit** [comment]
Hello, I wonder if this is appropriate. I have attempted to setup the KLC of colemak with no success, I am currently using EPKL for Colemak-dh - it works wonderfully well! After fiddling for far too long with MSKLC I got nowhere, I explored (as best as I could) the suggestions that you have made in the MSKLC folder of the BigBag.... it went nowhere too.  
I was unable to get any dead keys to work as I desired [press " +  spacebar] and only functions as ["+symbol/letter], quite helpless here. 

Could it be that there isn't a functioning .klc file for colemak anywhere? I doubt that could be possible... I wouldn't even know how to check.
I have 3 questions: 
Is MSKLC viable for colemak (I failed to get the .klc files to run)? 
What is the "APT" layout in EPKL please?
Any suggestions on how to get the deadkeys to work please?

**Reply by DreymaR** [reply]
There are several functioning .klc files around.

One common beginner problem is downloading the file wrong so you don't get line endings (need to be Windows CRLF) and/or encoding (need to be whatever MSKLC uses; an UTF-16 variant iirc). This is described in the EPKL MSKLC readme (in the Other\MSKLC folder).

**Reply by wismill** [reply]
You may be interested in [kalamine](https://github.com/OneDeadKey/kalamine). It really easy to design a layout in ASCII art. As for MSKLC, it fixes some of its bugs/limitations by tweaking the compilation of the driver.

**Reply by DaniilHubGit** [reply]
I will look into this soon, thank you for such the quick reply!

**Reply by DaniilHubGit** [reply]
.klc file for colemak dh works. All that was needed was patience and to restart my system, no need to fiddle around with anything else. Thank you for the support :)

**Reply by DreymaR** [reply]
Oh, I see. Yes, not restarting the system after an install must be the most common MSKLC beginner's mistake. These days most programs don't require you to think about restarting as they either just up and run or explicitly ask the system for a restart. MSKLC is an older program and Microsoft doesn't seem to assign any resources to keeping it modern, so users need to understand some old-school practices to use it. I did mention this in the MSKLC Readme.

**Reply by DaniilHubGit** [reply]
The instructions are in bold :P, should have seen that.
